editoast/simulation/src/lib.rs

```diff
@@ -1,3 +1,341 @@
+         *                                                |                                  |           ^                        |
+         *                                                |                                  |           |(cache miss)            v
+         *                                                |                                  |
+         * requests stream ----------> [cache_read] ------+-------> [chunk_processing] ------+----> [aggregation]---------> result_stream
+         * (self, function                                |                                  |                   (cache hit)
+         *     input)                                     |                                  |
+         *                                                |                                  |
+         *                                                +--------> ... --------------------+
+         *                                  (spawns several tasks
+         *                                   via for_each_concurrent)
+         */
+
+        // shared to several tasks
+        let vkconn = Arc::new(tokio::sync::Mutex::new(vkconn));
+
+        let (cache_write_tx, mut cache_write_rx) =
+            tokio::sync::mpsc::unbounded_channel::<(String, String)>();
+        {
+            let vkconn = vkconn.clone();
+            // 'write_cache' task, writes input key-value pairs to cache, logging errors
+            tokio::spawn(
+                async move {
+                    use deadpool_redis::redis::AsyncCommands as _;
+                    while let Some((key, serialized_value)) = cache_write_rx.recv().await {
+                        if let Err(e) = vkconn
+                            .lock()
+                            .await
+                            .set::<_, _, ()>(key, serialized_value)
+                            .await
+                        {
+                            tracing::error!(?e, "simulation stream: cache write failure")
+                        }
+                    }
+                }
+                .in_current_span(),
+            );
+        }
+
+        let (cache_read_tx, mut cache_read_rx) = tokio::sync::mpsc::unbounded_channel::<(
+            T, // input
+            CorrelationKey,
+            String,            // cache key
+            Option<T::Output>, // maybe a cached simulation
+        )>();
+        {
+            // 'cache_read' task, consumes input stream (self), chunks cache reads,
+            // processes each chunk in a dedicated task 'chunk_processing' (spawned by for_each_concurrent),
+            // and sends a bunch of data to the 'aggregation' task
+            tokio::spawn(
+                self.chunks(T::CACHE_READS_BATCH_SIZE)
+                    .zip(stream::iter(iter::repeat(vkconn.clone())))
+                    .zip(stream::iter(iter::repeat(cache_read_tx)))
+                    .for_each_concurrent(None, async move |((inputs, vkconn), tx)| {
```

Comment:
minor: I'd keep the full name to understand it better later in this big closure.

```suggestion
                    .for_each_concurrent(None, async move |((inputs, vkconn), cache_read_tx)| {
```

Reply:
agreed 7066ad7d0a